server/service/users.go

```diff
@@ -0,0 +1,521 @@
+}
+
+func (svc *Service) CreateUser(ctx context.Context, p fleet.UserPayload) (*fleet.User, error) {
+	var teams []fleet.UserTeam
+	if p.Teams != nil {
+		teams = *p.Teams
+	}
+	if err := svc.authz.Authorize(ctx, &fleet.User{Teams: teams}, fleet.ActionWrite); err != nil {
+		return nil, err
+	}
+
+	if invite, err := svc.ds.InviteByEmail(ctx, *p.Email); err == nil && invite != nil {
+		return nil, ctxerr.Errorf(ctx, "%s already invited", *p.Email)
+	}
+
+	if p.AdminForcedPasswordReset == nil {
+		// By default, force password reset for users created this way.
+		p.AdminForcedPasswordReset = ptr.Bool(true)
+	}
+
+	return svc.newUser(ctx, p)
+}
+
+////////////////////////////////////////////////////////////////////////////////
+// List Users
+////////////////////////////////////////////////////////////////////////////////
+
+type listUsersRequest struct {
+	ListOptions fleet.UserListOptions `url:"user_options"`
+}
+
+type listUsersResponse struct {
+	Users []fleet.User `json:"users"`
+	Err   error        `json:"error,omitempty"`
+}
+
+func (r listUsersResponse) error() error { return r.Err }
+
+func listUsersEndpoint(ctx context.Context, request interface{}, svc fleet.Service) (interface{}, error) {
+	req := request.(*listUsersRequest)
+	users, err := svc.ListUsers(ctx, req.ListOptions)
+	if err != nil {
+		return listUsersResponse{Err: err}, nil
+	}
+
+	resp := listUsersResponse{Users: []fleet.User{}}
+	for _, user := range users {
+		resp.Users = append(resp.Users, *user)
+	}
+	return resp, nil
+}
+
+func (svc *Service) ListUsers(ctx context.Context, opt fleet.UserListOptions) ([]*fleet.User, error) {
+	if err := svc.authz.Authorize(ctx, &fleet.User{}, fleet.ActionRead); err != nil {
+		return nil, err
+	}
+
+	return svc.ds.ListUsers(ctx, opt)
+}
+
+////////////////////////////////////////////////////////////////////////////////
+// Get User
+////////////////////////////////////////////////////////////////////////////////
+
+type getUserRequest struct {
+	ID uint `url:"id"`
+}
+
+type getUserResponse struct {
+	User *fleet.User `json:"user,omitempty"`
+	Err  error       `json:"error,omitempty"`
+}
+
+func (r getUserResponse) error() error { return r.Err }
+
+func getUserEndpoint(ctx context.Context, request interface{}, svc fleet.Service) (interface{}, error) {
+	req := request.(*getUserRequest)
+	user, err := svc.User(ctx, req.ID)
+	if err != nil {
+		return getUserResponse{Err: err}, nil
+	}
+	return getUserResponse{User: user}, nil
+}
+
+func (svc *Service) User(ctx context.Context, id uint) (*fleet.User, error) {
+	if err := svc.authz.Authorize(ctx, &fleet.User{ID: id}, fleet.ActionRead); err != nil {
+		return nil, err
+	}
+
+	return svc.ds.UserByID(ctx, id)
+}
+
+////////////////////////////////////////////////////////////////////////////////
+// Modify User
+////////////////////////////////////////////////////////////////////////////////
+
+type modifyUserRequest struct {
+	ID uint `json:"-" url:"id"`
```

Comment:
Do we need the `json:"-"`? (e.g. `getUserRequest ` doesn't set it)

Reply:
For now, yes, because the whole body is JSON-decoded into the struct but the ID comes from the URL (it must not get overwritten by decoding the body). However the goal is to eventually not have to add that tag on the request struct (e.g. when decoding json in the struct, automatically consider those with non-json tags to be left alone by json). That will come in a future PR though.